File: examples/python/ledger.py
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""`LedgerEntry` mirrors examples/cases/ledger.yaml.

`@serify_model` is the entire schema binding — it reads the dataclass annotations
and generates `from_field_map` / `to_field_map`, so nothing here calls a
`get_*`/`set_*` accessor. Everything else is the byte layout, which is the part a
conformance worker exists to exercise.

Go is the --ref language and owns that layout; see the comment at the top of
examples/go/wire.go.
"""

import struct
from dataclasses import dataclass

from serify import serify_model

INT128_MASK = (1 << 128) - 1


@serify_model
@dataclass
class LedgerEntry:
    entry_id: int
    block_number: int
    block_time: int
    tx_hash: bytes
    account: str
    asset: str
    amount_base_units: int
    balance_after: int
    confirmed: bool
    memo: "str | None"

    def marshal(self) -> bytes:
        buf = bytearray()

        buf += struct.pack("<QQq", self.entry_id, self.block_number, self.block_time)
        buf += struct.pack("<I", len(self.tx_hash)) + self.tx_hash

        for s in (self.account, self.asset):
            b = s.encode()
            buf += struct.pack("<I", len(b)) + b

        # int128: 16 bytes little-endian, two's complement. Masking to 128 bits
        # maps a negative onto its residue class, which is exactly two's complement.
        for v in (self.amount_base_units, self.balance_after):
            buf += (v & INT128_MASK).to_bytes(16, "little")

        buf += struct.pack("<?", self.confirmed)

        if self.memo is None:
            buf += b"\x00"
        else:
            m = self.memo.encode()
            buf += b"\x01" + struct.pack("<I", len(m)) + m

        return bytes(buf)

    @classmethod
    def unmarshal(cls, data: bytes) -> "LedgerEntry":
        off = 0
        entry_id, block_number, block_time = struct.unpack_from("<QQq", data, off)
        off += 24

        (n,) = struct.unpack_from("<I", data, off)
        off += 4
        tx_hash = data[off:off + n]
        off += n

        strs = []
        for _ in range(2):
            (n,) = struct.unpack_from("<I", data, off)
            off += 4
            strs.append(data[off:off + n].decode())
            off += n

        ints = []
        for _ in range(2):
            v = int.from_bytes(data[off:off + 16], "little")
            off += 16
            if v >= 1 << 127:  # re-apply the sign
                v -= 1 << 128
            ints.append(v)

        (confirmed,) = struct.unpack_from("<?", data, off)
        off += 1

        memo = None
        if data[off]:
            off += 1
            (n,) = struct.unpack_from("<I", data, off)
            off += 4
            memo = data[off:off + n].decode()

        return cls(
            entry_id=entry_id, block_number=block_number, block_time=block_time,
            tx_hash=tx_hash, account=strs[0], asset=strs[1],
            amount_base_units=ints[0], balance_after=ints[1],
            confirmed=confirmed, memo=memo,
        )

File: examples/node/src/ledger.ts
/**
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * `LedgerEntry` mirrors examples/cases/ledger.yaml.
 *
 * `@Serify.Model()` plus one `@Serify.field()` per property is the entire schema
 * binding — nothing here calls a get/set accessor. Everything else is the byte
 * layout, which is the part a conformance worker exists to exercise.
 *
 * JS numbers are IEEE-754 doubles, so every 64/128-bit integer is a BigInt.
 * BigInt is unbounded, so the two int128 fields need no special type — but
 * Buffer has no 128-bit writer, so they are written as two 64-bit halves.
 *
 * Go is the --ref language and owns the layout; see examples/go/wire.go.
 */

import { Serify } from '@chengxilo/serify';

import { lenPrefixed, lenPrefixedStr } from './wire';

const MASK_64 = (1n << 64n) - 1n;
const MASK_128 = (1n << 128n) - 1n;

// int128: 16 bytes little-endian, two's complement. Masking to 128 bits maps a
// negative onto its residue class, which is exactly two's complement.
function writeInt128LE(v: bigint): Buffer {
  const u = v & MASK_128;
  const b = Buffer.alloc(16);
  b.writeBigUInt64LE(u & MASK_64, 0);
  b.writeBigUInt64LE(u >> 64n, 8);
  return b;
}

function readInt128LE(b: Buffer, off: number): bigint {
  const u = b.readBigUInt64LE(off) | (b.readBigUInt64LE(off + 8) << 64n);
  return u >= 1n << 127n ? u - (1n << 128n) : u; // re-apply the sign
}

@Serify.Model()
export class LedgerEntry {
  @Serify.field() entry_id: bigint = 0n;
  @Serify.field() block_number: bigint = 0n;
  @Serify.field() block_time: bigint = 0n;
  @Serify.field() tx_hash: Buffer = Buffer.alloc(0);
  @Serify.field() account = '';
  @Serify.field() asset = '';
  @Serify.field() amount_base_units: bigint = 0n;
  @Serify.field() balance_after: bigint = 0n;
  @Serify.field() confirmed = false;
  @Serify.field() memo: string | null = null;

  marshal(): Buffer {
    const head = Buffer.alloc(24);
    head.writeBigUInt64LE(this.entry_id, 0);
    head.writeBigUInt64LE(this.block_number, 8);
    head.writeBigInt64LE(this.block_time, 16);

    const parts: Buffer[] = [
      head,
      lenPrefixed(this.tx_hash),
      lenPrefixedStr(this.account),
      lenPrefixedStr(this.asset),
      writeInt128LE(this.amount_base_units),
      writeInt128LE(this.balance_after),
      Buffer.from([this.confirmed ? 1 : 0]),
    ];

    if (this.memo === null) {
      parts.push(Buffer.from([0]));
    } else {
      parts.push(Buffer.from([1]), lenPrefixedStr(this.memo));
    }
    return Buffer.concat(parts);
  }

  static unmarshal(data: Buffer): LedgerEntry {
    const e = new LedgerEntry();
    e.entry_id = data.readBigUInt64LE(0);
    e.block_number = data.readBigUInt64LE(8);
    e.block_time = data.readBigInt64LE(16);
    let off = 24;

    const readLenBuf = (): Buffer => {
      const n = data.readUInt32LE(off);
      off += 4;
      const b = data.subarray(off, off + n);
      off += n;
      return Buffer.from(b); // copy: a subarray would alias the input
    };

    e.tx_hash = readLenBuf();
    e.account = readLenBuf().toString('utf8');
    e.asset = readLenBuf().toString('utf8');

    e.amount_base_units = readInt128LE(data, off);
    e.balance_after = readInt128LE(data, off + 16);
    off += 32;

    e.confirmed = data[off] !== 0;
    const hasMemo = data[off + 1] !== 0;
    off += 2;
    e.memo = hasMemo ? readLenBuf().toString('utf8') : null;

    return e;
  }
}
